File: ios/Sasquatsh/ViewModels/GroupListViewModel.swift
import SwiftUI

@Observable
@MainActor
final class GroupListViewModel {
    var groups: [GroupSummary] = []
    var myGroupCount = 0
    var isLoading = false
    var error: String?
    var searchText = ""
    var selectedType: GroupType?
    var filterCity = ""
    var filterState = ""

    private var services: ServiceContainer?

    func configure(services: ServiceContainer) {
        self.services = services
    }

    var filter: GroupSearchFilter {
        GroupSearchFilter(
            search: searchText.isEmpty ? nil : searchText,
            groupType: selectedType,
            city: filterCity.isEmpty ? nil : filterCity,
            state: filterState.isEmpty ? nil : filterState
        )
    }

    var hasActiveFilters: Bool {
        selectedType != nil || !filterCity.isEmpty || !filterState.isEmpty
    }

    func loadGroups() async {
        guard let services else { return }
        isLoading = true
        error = nil
        do {
            async let publicGroups = services.groups.getPublicGroups(filter: filter)
            async let myGroupsList = services.groups.getMyGroups()

            var allGroups = try await publicGroups
            let myGroups = (try? await myGroupsList) ?? []
            myGroupCount = myGroups.count

            // Build a map of group ID → user role from my groups
            let roleMap = Dictionary(uniqueKeysWithValues: myGroups.compactMap { g -> (String, MemberRole)? in
                guard let role = g.userRole else { return nil }
                return (g.id, role)
            })

            // Annotate public groups with user's role
            for i in allGroups.indices {
                if let role = roleMap[allGroups[i].id] {
                    allGroups[i].userRole = role
                }
            }

            // Sort: user's groups first (by role: owner > admin > member), then others
            groups = allGroups.sorted { a, b in
                let aRole = a.userRole
                let bRole = b.userRole
                if aRole != nil && bRole == nil { return true }
                if aRole == nil && bRole != nil { return false }
                if let ar = aRole, let br = bRole {
                    let order: [MemberRole: Int] = [.owner: 0, .admin: 1, .member: 2]
                    return (order[ar] ?? 3) < (order[br] ?? 3)
                }
                return a.name < b.name
            }
        } catch {
            self.error = error.localizedDescription
        }
        isLoading = false
    }

    func clearFilters() {
        searchText = ""
        selectedType = nil
        filterCity = ""
        filterState = ""
    }
}

File: android/app/src/main/java/com/sasquatsh/app/viewmodels/GroupListViewModel.kt
package com.sasquatsh.app.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.sasquatsh.app.models.GroupSearchFilter
import com.sasquatsh.app.models.GroupSummary
import com.sasquatsh.app.models.GroupType
import com.sasquatsh.app.models.MemberRole
import com.sasquatsh.app.services.GroupsService
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.async
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

data class GroupListUiState(
    val groups: List<GroupSummary> = emptyList(),
    val myGroupCount: Int = 0,
    val isLoading: Boolean = false,
    val error: String? = null,
    val searchText: String = "",
    val selectedType: GroupType? = null,
    val filterCity: String = "",
    val filterState: String = ""
) {
    val hasActiveFilters: Boolean
        get() = selectedType != null || filterCity.isNotEmpty() || filterState.isNotEmpty()

    val filter: GroupSearchFilter
        get() = GroupSearchFilter(
            search = searchText.ifEmpty { null },
            groupType = selectedType,
            city = filterCity.ifEmpty { null },
            state = filterState.ifEmpty { null }
        )
}

@HiltViewModel
class GroupListViewModel @Inject constructor(
    private val groupsService: GroupsService
) : ViewModel() {

    private val _uiState = MutableStateFlow(GroupListUiState())
    val uiState: StateFlow<GroupListUiState> = _uiState.asStateFlow()

    fun updateSearchText(text: String) {
        _uiState.update { it.copy(searchText = text) }
    }

    fun updateSelectedType(type: GroupType?) {
        _uiState.update { it.copy(selectedType = type) }
    }

    fun updateFilterCity(city: String) {
        _uiState.update { it.copy(filterCity = city) }
    }

    fun updateFilterState(state: String) {
        _uiState.update { it.copy(filterState = state) }
    }

    fun loadGroups() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, error = null) }
            try {
                val publicGroupsDeferred = async { groupsService.getPublicGroups(_uiState.value.filter) }
                val myGroupsDeferred = async {
                    try {
                        groupsService.getMyGroups()
                    } catch (_: Exception) {
                        emptyList()
                    }
                }

                val publicGroups = publicGroupsDeferred.await()
                val myGroups = myGroupsDeferred.await()

                // Build a map of group ID -> user's role from my groups
                val roleMap = myGroups.associate { it.id to it.userRole }

                // Annotate public groups with user's role, dedup against my groups
                val myGroupIds = myGroups.map { it.id }.toSet()
                val annotatedPublic = publicGroups
                    .filter { it.id !in myGroupIds }
                    .map { group ->
                        val role = roleMap[group.id]
                        if (role != null) group.copy(userRole = role) else group
                    }

                // Merge: my groups first (sorted by role priority), then other groups alphabetically
                val rolePriority = mapOf(
                    MemberRole.OWNER to 0,
                    MemberRole.ADMIN to 1,
                    MemberRole.MEMBER to 2
                )
                val sortedMyGroups = myGroups.sortedWith(
                    compareBy<GroupSummary> { rolePriority[it.userRole] ?: 3 }
                        .thenBy { it.name.lowercase() }
                )
                val sortedOtherGroups = annotatedPublic.sortedBy { it.name.lowercase() }
                val merged = sortedMyGroups + sortedOtherGroups

                _uiState.update {
                    it.copy(
                        groups = merged,
                        myGroupCount = myGroups.size,
                        isLoading = false
                    )
                }
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.localizedMessage, isLoading = false) }
            }
        }
    }

    fun clearFilters() {
        _uiState.update {
            it.copy(
                searchText = "",
                selectedType = null,
                filterCity = "",
                filterState = ""
            )
        }
    }
}
